test/suitesOrder.js

```diff
@@ -0,0 +1,157 @@
+var execHistoryCopy,
+	execHistory = [];
+
+(function( suite, test, beforeEach, afterEach ) {
+
+	beforeEach(function( assert ) {
+		execHistory.push( assert.suite.fullName + ".beforeEach1" );
+	});
+	beforeEach(function( assert ) {
+		execHistory.push( assert.suite.fullName + ".beforeEach2" );
+	});
+
+	test( "test1", function( assert ) {
+		assert.expect( 0 );
+		execHistory.push( assert.suite.fullName + " -> test1.callback" );
+	});
+
+	suite( "level1", function() {
+		beforeEach(function( assert ) {
+			execHistory.push( assert.suite.fullName + ".beforeEach1" );
+		});
+		beforeEach(function( assert ) {
+			execHistory.push( assert.suite.fullName + ".beforeEach2" );
+		});
+
+		test( "test2", function( assert ) {
+			assert.expect( 0 );
+			execHistory.push( assert.suite.fullName + " -> test2.callback" );
+		});
+
+		suite( "level2", function() {
```

Comment:
Tying invocation of global `beforeEach`/`afterEach`/`test`/`suite` functions to suite context by virtue of being executed during a suite callback feels dirty (and _exactly_ like tying invocation of global assertion functions to test context by virtue of being executed during a test callback, which is deservedly on the way out). I know there are backcompat reasons for that behavior with `module`, but I think we can do better with this `suite` green field.

I'd like them to instead be accessible from either context or arguments, which directly relates to (and partly refutes) the end of https://github.com/jquery/qunit/pull/670#issuecomment-78513676 . _How_ to do so may be a matter of great bikeshedding, though. Some of my own goals on that point are as follows, in no particular order:
* Intuitively-readable invocations, possibly paradigm-dependent (e.g., use of BDD names)
* Maximum similarity between `suite` and `test` callbacks, for future iterative convergence
* Separate concerns (e.g., no properties on `assert` that do not pertain to assertions, since we might want to allow integrating external assertion libraries)